File: day4.c
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#define A(i, j) a[i + 4][j + 4]
char a[256][256];
int n;

bool match(const char* m, int f, int i, int j, int di, int dj)
{
    for (int k = 0; k < strlen(m); k++) {
        int s = i + di * k;
        int t = j + dj * k;
        if (f & 1) s = n - s - 1;
        if (f & 2) t = n - t - 1;
        if (f & 4) { int u = s; s = t; t = u; }
        if (A(s, t) != m[k]) return false;
    }
    return true;
}

int main()
{
    FILE* file = fopen("day4.txt", "r");
    int sum = 0, sum2 = 0;

    while (1 == fscanf(file, "%s", &A(n++, 0))) ;

    for (int f = 0; f < 8; f++)
    for (int i = 0; i < n; i++)
    for (int j = 0; j < n; j++) {
        if (match("XMAS", f, i, j, 1, 0)) sum++;
        if (match("XMAS", f, i, j, 1, 1)) sum++;
        if (match("MAS", f, i, j, 1, 1) && 
            match("MAS", f, i, j + 2, 1, -1)) sum2++;
    }

    printf("%d %d\n", sum / 2, sum2 / 2);
}

File: day4.lua
#!/usr/bin/env lua

local a = {}

local function match(m, f, i, j, di, dj)
    for k = 1, #m do
        local s = i + di * (k - 1)
        local t = j + dj * (k - 1)
        if f & 1 ~= 0 then s = #a - s + 1 end
        if f & 2 ~= 0 then t = #a - t + 1 end
        if f & 4 ~= 0 then s, t = t, s end
        if (a[s] or ""):byte(t) ~= m:byte(k) then
            return false
        end
    end
    return true
end

local file = io.open("day4.txt")
local sum, sum2 = 0, 0

for line in file:lines() do
    table.insert(a, line)
end

for f = 0, 7 do
    for i = 1, #a do
        for j = 1, #a do
            if match("XMAS", f, i, j, 1, 0) then
                sum = sum + 1
            end
            if match("XMAS", f, i, j, 1, 1) then
                sum = sum + 1
            end
            if match("MAS", f, i, j, 1, 1) and
               match("MAS", f, i, j + 2, 1, -1) then
                sum2 = sum2 + 1
            end
        end
    end
end

print(sum / 2, sum2 / 2)
